File: FastGenerator/Generator.cpp
#include "generator.h"

using namespace std;

generator::generator(char** str_array, const int size)
{
	size_ = size;
	words_ = str_array;
}

list<char*> generator::check_input(const char* input) const
{
	list<char*> word_list;
	for (auto i = 0; i < size_; i++)
	{
		auto word = words_[i];
		
		if (check_word(input, word))
		{
			word_list.push_back(word);
		}
	}
	return word_list;
}

bool generator::check_word(const char* input, const char* word_to_check)
{
	list<char> sequence;
	for (auto i = 0; input[i] != 0; i++)
		sequence.push_back(input[i]);
	
	for (auto i = 0; word_to_check[i] != 0; i++)
	{
		auto element = find(sequence.begin(), sequence.end(), word_to_check[i]);
		
		if (element == sequence.end())
			return false;

		sequence.erase(element);
	}
	return true;
}

File: Model/Generator.cs
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;

namespace Model
{
    public class Generator : IGenerator
    {
        private readonly List<string> _words = new List<string>();

        public Generator()
        {
            LoadWords();
        }

        private void LoadWords()
        {
            using (var fs = File.Open(@"words.txt", FileMode.Open, FileAccess.Read))
            using (var bs = new BufferedStream(fs))
            using (var sr = new StreamReader(bs))
            {
                string line;
                while ((line = sr.ReadLine()) != null)
                {
                    _words.Add(line);
                }
            }
        }

        public IList CheckInput(string input)
        {
            return _words.Where(word => CheckWord(input, word)).ToList();
        }

        private static bool CheckWord(string input, string wordToCheck)
        {
            var sequence = input.ToCharArray().ToList();
            foreach (var character in wordToCheck)
            {
                if (!sequence.Contains(character))
                    return false;
                sequence.Remove(character);
            }
            return true;
        }

    }
}
